src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs

```diff
@@ -273,42 +296,42 @@ private void OnEventSourceCommand(object sender, EventCommandEventArgs e)
 
         #region Global EventCounterGroup Array management
 
-        private static EventCounterGroup[] s_eventCounterGroups;
-
-        internal static void AddEventCounter(EventSource eventSource, EventCounter eventCounter)
-        {
-            int eventSourceIndex = EventListenerHelper.EventSourceIndex(eventSource);
-
-            EventCounterGroup.EnsureEventSourceIndexAvailable(eventSourceIndex);
-            EventCounterGroup eventCounterGroup = GetEventCounterGroup(eventSource);
-            eventCounterGroup.Add(eventCounter);
-        }
+        // We need eventCounters to 'attach' themselves to a particular EventSource.   
+        // this table provides the mapping from EventSource -> EventCounterGroup 
+        // which represents this 'attached' information.   
+        private static WeakReference<EventCounterGroup>[] s_eventCounterGroups;
+        private static object s_eventCounterGroupsLock = new object();
 
         private static void EnsureEventSourceIndexAvailable(int eventSourceIndex)
         {
+            Debug.Assert(Monitor.IsEntered(s_eventCounterGroupsLock));
             if (EventCounterGroup.s_eventCounterGroups == null)
             {
-                EventCounterGroup.s_eventCounterGroups = new EventCounterGroup[eventSourceIndex + 1];
+                EventCounterGroup.s_eventCounterGroups = new WeakReference<EventCounterGroup>[eventSourceIndex + 1];
             }
             else if (eventSourceIndex >= EventCounterGroup.s_eventCounterGroups.Length)
             {
-                EventCounterGroup[] newEventCounterGroups = new EventCounterGroup[eventSourceIndex + 1];
+                WeakReference<EventCounterGroup>[] newEventCounterGroups = new WeakReference<EventCounterGroup>[eventSourceIndex + 1];
                 Array.Copy(EventCounterGroup.s_eventCounterGroups, 0, newEventCounterGroups, 0, EventCounterGroup.s_eventCounterGroups.Length);
                 EventCounterGroup.s_eventCounterGroups = newEventCounterGroups;
             }
         }
 
-        private static EventCounterGroup GetEventCounterGroup(EventSource eventSource)
+        internal static EventCounterGroup GetEventCounterGroup(EventSource eventSource)
         {
-            int eventSourceIndex = EventListenerHelper.EventSourceIndex(eventSource);
-            EventCounterGroup result = EventCounterGroup.s_eventCounterGroups[eventSourceIndex];
-            if (result == null)
+            lock (s_eventCounterGroupsLock)
             {
-                result = new EventCounterGroup(eventSource, eventSourceIndex);
-                EventCounterGroup.s_eventCounterGroups[eventSourceIndex] = result;
+                int eventSourceIndex = EventListenerHelper.EventSourceIndex(eventSource);
+                EnsureEventSourceIndexAvailable(eventSourceIndex);
+                var weakRef = EventCounterGroup.s_eventCounterGroups[eventSourceIndex];
```

Comment:
Nit: we try to use var only when the type is clear from the right-hand side, which it's not in this case.